### PR TITLE
Support opaque pointers in ReplacePointerBitcast and InlineFuncWithPointerBitCastArg

### PR DESCRIPTION
**Pull request by alan-baker**
Contributes to #816

* Inline functions where there is an implicit pointer cast in the
  arguments
* Support implicit pointer casts in replace pointer bitcasts
* Cover alloca and global variables for inferring types from the result
* Remove some unnecessary pointer casts in ReplaceOpenCLBuiltin when
  using opaque pointers

### CODE REVIEW

**Comment by alan-baker** [comment]
Worth noting that both major modifications are basically just identifying implicit pointer casts because the bitcast instructions should no longer appear with opaque pointers.

File: lib/InlineFuncWithPointerBitCastArgPass.cpp [hunk 1]
```diff
@@ -93,6 +94,86 @@ bool clspv::InlineFuncWithPointerBitCastArgPass::InlineFunctions(Module &M) {
...
+                to_check.push_back(
+                    std::make_pair(use.getUser(), use.getOperandNo()));
+              }
+              break;
+            case Instruction::GetElementPtr: {
+              auto *next_gep = cast<GetElementPtrInst>(user);
+              if (result_ele_ty &&
+                  next_gep->getResultElementType() != result_ele_ty) {
+                // TODO: this is an over-approximation, but it's easier to not
+                // worry about phis.
```
Comment (dneto0) [inline lib/InlineFuncWithPointerBitCastArgPass.cpp:161]: nit: This is a funny comment.  It talks about Phis, but this is the GEP case.

File: lib/ReplacePointerBitcastPass.cpp [hunk 2]
```diff
@@ -522,8 +594,16 @@ clspv::ReplacePointerBitcastPass::run(Module &M, ModuleAnalysisManager &) {
...
+    // TODO(#816): remove after final transition.
+    Value *start = isa<BitCastInst>(Inst) ? Inst : Src;
+    for (User *BitCastUser : start->users()) {
+      if (ImplicitCasts.count(BitCastUser)) {
+        // If this user was queued on the worklist as an implicit cast
+        // separately, don't handle it now.
+        continue;
+      }
+
+     LLVM_DEBUG(dbgs() << "#### BitCastUser: "; BitCastUser->dump());
```
Comment (dneto0) [inline lib/ReplacePointerBitcastPass.cpp:606]: nit: indent one space :-)

**Review by dneto0 — APPROVED** [review]
Thanks!

**Pull request merged**